packages/x-data-grid-generator/src/hooks/serverUtils.ts

```diff
@@ -0,0 +1,492 @@
+): Promise<FakeServerResponse> => {
+  const { minDelay = 100, maxDelay = 300, useCursorPagination } = serverOptions;
+
+  if (maxDelay < minDelay) {
+    throw new Error('serverOptions.minDelay is larger than serverOptions.maxDelay ');
+  }
+  const delay = randomInt(minDelay, maxDelay);
+
+  const { cursor, page = 0, pageSize } = queryOptions;
+
+  let nextCursor;
+  let firstRowIndex;
+  let lastRowIndex;
+
+  let filteredRows = getFilteredRows(rows, queryOptions.filterModel, columnsWithDefaultColDef);
+
+  const rowComparator = getRowComparator(queryOptions.sortModel, columnsWithDefaultColDef);
+  filteredRows = [...filteredRows].sort(rowComparator);
+
+  const totalRowCount = filteredRows.length;
+  if (!pageSize) {
+    firstRowIndex = 0;
+    lastRowIndex = filteredRows.length;
+  } else if (useCursorPagination) {
+    firstRowIndex = cursor ? filteredRows.findIndex(({ id }) => id === cursor) : 0;
+    firstRowIndex = Math.max(firstRowIndex, 0); // if cursor not found return 0
+    lastRowIndex = firstRowIndex + pageSize;
+
+    nextCursor = lastRowIndex >= filteredRows.length ? undefined : filteredRows[lastRowIndex].id;
+  } else {
+    firstRowIndex = page * pageSize;
+    lastRowIndex = (page + 1) * pageSize;
+  }
+  const hasNextPage = lastRowIndex < filteredRows.length - 1;
+  const response: FakeServerResponse = {
+    returnedRows: filteredRows.slice(firstRowIndex, lastRowIndex),
+    hasNextPage,
+    nextCursor,
+    totalRowCount,
+  };
+
+  return new Promise<FakeServerResponse>((resolve) => {
+    setTimeout(() => {
+      resolve(response);
+    }, delay); // simulate network latency
+  });
+};
+
+interface ProcessTreeDataRowsResponse {
+  rows: GridRowModel[];
+  rootRowCount: number;
+}
+
+const findTreeDataRowChildren = (
+  allRows: GridRowModel[],
+  parentPath: string[],
+  pathKey: string = 'path',
+  depth: number = 1, // the depth of the children to find relative to parentDepth, `-1` to find all
+) => {
+  const parentDepth = parentPath.length;
+  const children = [];
+  for (let i = 0; i < allRows.length; i += 1) {
+    const row = allRows[i];
+    const rowPath = row[pathKey];
+    if (!rowPath) {
+      continue;
+    }
+    if (
+      ((depth < 0 && rowPath.length > parentDepth) || rowPath.length === parentDepth + depth) &&
+      parentPath.every((value, index) => value === rowPath[index])
+    ) {
+      children.push(row);
+    }
+  }
+  return children;
+};
+
+type GetTreeDataFilteredRows = (
+  rows: GridValidRowModel[],
+  filterModel: GridFilterModel | undefined,
+  columnsWithDefaultColDef: GridColDef[],
+) => GridValidRowModel;
+
+const getTreeDataFilteredRows: GetTreeDataFilteredRows = (
+  rows,
+  filterModel,
+  columnsWithDefaultColDef,
+): GridValidRowModel[] => {
+  let filteredRows = [...rows];
+  if (filterModel && filterModel.quickFilterValues?.length! > 0) {
+    filteredRows = getQuicklyFilteredRows(rows, filterModel, columnsWithDefaultColDef);
+  }
+  if (filterModel?.items.length === 0) {
```

Comment:
```suggestion
  if (filterModel?.items.length > 0) {
```

Otherwise filtering does not work